geps/gep-1713/index.md

```diff
@@ -0,0 +1,685 @@
+There can be O(1000) Knative `Services` in the cluster which means we have O(1000) distinct certificates.
+Thus updating a single `Gateway` resource with this many certificates is a contention point and inhibits horizontal scaling of our controllers.
+[Istio Ambient](https://istio.io/v1.15/blog/2022/introducing-ambient-mesh/), similarly, creates a listener per Kubernetes service.
+
+More broadly, large scale gateway users often expose `O(1000)` domains, but are currently limited by the maximum of 64 `listeners`.
+
+The spec currently has language to indicate implementations `MAY` merge `Gateways` resources but the mechanic isn't defined.
+https://github.com/kubernetes-sigs/gateway-api/blob/541e9fc2b3c2f62915cb58dc0ee5e43e4096b3e2/apis/v1beta1/gateway_types.go#L76-L78
+
+## Feature Details
+
+We define `ListenerSet` as the name of the feature outlined in this GEP.
+The feature will be part of the experimental branches as an extended feature, which implementations can choose to support. At that time, all the requirements in this document that use MUST apply to their implementation of the feature.
+
+
+## API
+
+This proposal introduces a new `ListenerSet` resource that has the ability to attach to a set of listeners to multiple parent `Gateways`.
+
+### Go
+
+```go
+type GatewaySpec struct {
+	...
+	// Note: this is a list to allow future potential features
+	AllowedListeners []*AllowedListeners `json:"allowedListeners"`
+	...
+}
+
+type AllowedListeners struct {
+	// +kubebuilder:default={from: None}
+	Namespaces *ListenerNamespaces `json:"namespaces,omitempty"`
+}
+
+// ListenerNamespaces indicate which namespaces ListenerSets should be selected from.
+type ListenerNamespaces struct {
+	// From indicates where ListenerSets can attach to this Gateway. Possible
+	// values are:
+	//
+	// * Same: Only ListenerSets in the same namespace may be attached to this Gateway.
+	// * None: Only listeners defined in the Gateway's spec are allowed
+	//
+	// +optional
+	// +kubebuilder:default=Same
+	// +kubebuilder:validation:Enum=Same;None
+	From *FromNamespaces `json:"from,omitempty"`
+}
+
+// ListenerSet defines a set of additional listeners to attach to an existing Gateway.
+type ListenerSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ListenerSet.
+	Spec ListenerSetSpec `json:"spec"`
+
+	// Status defines the current state of ListenerSet.
+	Status ListenerSetStatus `json:"status,omitempty"`
+}
+
+// ListenerSetSpec defines the desired state of a ListenerSet.
+type ListenerSetSpec struct {
+	// ParentRefs references the Gateway that the listeners are attached to.
+	//
+	// +kubebuilder:validation:MaxItems=2
+	ParentRefs []ParentGatewayReference `json:"parentRefs,omitempty"`
+
+	// Listeners associated with this ListenerSet. Listeners define
+	// logical endpoints that are bound on this referenced parent Gateway's addresses.
+	//
```

Comment:
I think we'll need some reference to the fact that Listeners will be combined using the same rules as `Gateway.spec.Listeners`.

Reply:
updated the godoc